Product Block Editor: Map `SelectWithTextInput` and `DateTime` inputs to the custom product blocks

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

To be compatible with Product Block Editor, this PR makes the conversions for the last two attribute inputs:

- Implement the custom blocks for the product select-with-text field and product date-time field.
- Map `SelectWithTextInput` and `DateTime` inputs to the above custom product blocks.
- Set the minimum required Woo version of `AttributesBlock` class to 8.4 as it's the minimum version that the `__experimentalTextControl` component is exported from `@woocommerce/product-editor`.

### Screenshots:

#### 📷 Field is converted from text block to select-with-text block

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/c7295f35-294d-4df4-bf7a-edfbeb382561)

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/7f90e834-7e9f-4589-8931-4bd79e35510e)

#### 📷 Date-time block

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/4f96517a-bb89-4bbb-84f7-3a8e4810fc5a)

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/687873e2-16ce-420c-b0b9-dbacb6ee0d67)

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/7253359c-8054-404f-b74b-83ed4cca4604)

### Detailed test instructions:

#### 📌 Prepare test environment

1. Please refer to [the test preparation in PR 2151](https://github.com/woocommerce/google-listings-and-ads/pull/2151#prepare-test-environment) with **Woo 8.4.0**.
   - Currently, it can be tested with [8.4.0-rc.1](https://github.com/woocommerce/woocommerce/releases/tag/8.4.0-rc.1)
1. `npm install`
1. `npm start`

#### 📌 Test the select-with-text block

1. Install [WooCommerce Brands](https://github.com/woocommerce/woocommerce-brands) but don't activate it.
1. Go to edit a simple product, e.g. the **Beanie with Logo** imported from WC's sample products.
1. Switch to the **Organization** tab and find the **Google Listings & Ads** section.
1. The brand field should be shown in a text block.
1. Activate WooCommerce Brands and refresh the product editor page.
1. Check if the brand field is shown in a select-with-text block with three options:
   1. "Default"
   1. "From WooCommerce Brands"
   1. "Enter a custom value"
1. Change the selected option to see if the UI works well.
1. After selecting the "Enter a custom value" option, a text field should be shown below the selection.
1. Operate the selection and text fields.
1. Click on the **Update** button at the top-right of the page.
1. Refresh the page or switch back to the classical product editor to see if the changed option or text is saved.
1. The GTIN and MPN will be converted to use the select-with-text block when Yoast WooCommerce SEO is installed and activated.
   - If want to also test this part without actually getting the extension, an alternative way is to locally make the following function `return true;`
      https://github.com/woocommerce/google-listings-and-ads/blob/abd3492ab9db2350bfa7c215bdf452cb12b299ed/src/Integration/YoastWooCommerceSeo.php#L32-L34

#### 📌 Test the date-time block

1. Go to Settings > General. Set the site's timezone to another one different from your machine's system time, which is also the local timezone of browsers.
1. Go to edit a simple product that has not yet set an availability date.
1. Switch to the **Organization** tab and find the **Google Listings & Ads** section.
1. Check if there is an Availability Date field shown in a date-time block.
1. Operate the date-time block to set a date.
1. Click on the **Update** button at the top-right of the page.
1. Refresh the page to see if the availability date is saved.
1. Switch back to the classical product editor to see if the values on the availability date field are the same.
1. Set the site's timezone to another or the same as your machine.
1. Refresh and check both classical and blockified product editors to see if the date values are shifted as per the delta of the two timezones.

### Additional details:

TODOs will be cleaned up in a separate PR.

- https://github.com/woocommerce/google-listings-and-ads/blob/abd3492ab9db2350bfa7c215bdf452cb12b299ed/src/Admin/Input/Input.php#L60-L61
- https://github.com/woocommerce/google-listings-and-ads/blob/abd3492ab9db2350bfa7c215bdf452cb12b299ed/src/Admin/Product/Attributes/AttributesBlock.php#L194-L197

### Changelog entry

eason9487 committed:
Add the "Accessing product data" section to the README.md of custom product blocks.

eason9487 committed:
Implement the custom block for the product select-with-text field.

eason9487 committed:
Map `SelectWithTextInput` input to the custom block of the product select-with-text field.

eason9487 committed:
Set the minimum required Woo version of `AttributesBlock` to 8.4.

eason9487 committed:
Implement the custom block for the product date-time field.

eason9487 committed:
Map `DateTime` input to the custom block of the product date-time field.

puntope commented on js/src/blocks/README.md:
```diff
@@ -36,6 +36,24 @@ Since this extension requires a few custom blocks, and considering these custom
+- In block.json, it needs to list `"postType"` in the `usesContext` array to ask for the current post type ('product' or 'product_variation') from the context provider.
+- Forwarding the `context.postType` to `useProductEntityProp` hook is required in order to be compatible with both simple and variation product block templates.
+
+#### Derived value for initialization
```
❓  Can we elaborate or reword this explanation? Maybe my lack of knowledge of the concepts but not sure if I understand what those "derived values" are. I assume is just the data itself?

eason9487 replied:
In the React ecosystem, the derived value/state is commonly referred to the value calculated based on another value. I'm guessing that it probably comes from the wording in the [React API `getDerivedStateFromProps`](https://react.dev/reference/react/Component#static-getderivedstatefromprops).

Added in https://github.com/woocommerce/google-listings-and-ads/pull/2178/commits/0965ef92d88afe363b629b01caa45ecbacd23fa1.

puntope commented on js/src/blocks/index.js:
```diff
@@ -6,8 +6,12 @@ import { registerProductEditorBlockType } from '@woocommerce/product-editor';
  * Internal dependencies
  */
+import DatetimeFieldEdit from './product-date-time-field/edit';
+import datetimeFieldMetadata from './product-date-time-field/block.json';
```
I guess it should be `DateTimeFieldEdit` and `dateTimeFieldMetadata` because there is a `-` between date and time in the folder name.  Another option may be to call the folder `product-datetime-field`.

eason9487 replied:
Thanks! Fixed in https://github.com/woocommerce/google-listings-and-ads/pull/2178/commits/b94f33a402aed287545873ec7e144c8bae745bce.

puntope commented on js/src/blocks/product-date-time-field/block.json:
```diff
@@ -0,0 +1,30 @@
+	"$schema": "https://schemas.wp.org/trunk/block.json",
+	"apiVersion": 2,
+	"name": "google-listings-and-ads/product-date-time-field",
+	"title": "Product datetime control",
```
💅  Minor: can we keep it consistent with the block name conventions you are using?

So in this case:

"Product datetime field." 

 I don't mind control or field, but I think is better to call them in one single way. WDYT?

eason9487 replied:
Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/2178/commits/776ffb843618a06e08a985a84f56474ed3f49283.

puntope commented on js/src/blocks/product-date-time-field/edit.js:
💅  WDYT of renaming `value`, `setValue`, `setNextValue` to `dateTime`, `setDateTime`, `setNextDateTime`

So it's less generic.

eason9487 replied:
The use of `value` and `setValue` naming is to maintain consistency in the state and updater names that actually operate to the product data (wp-store) in each custom block.

puntope commented on js/src/blocks/product-date-time-field/edit.js:
⚠️  I found some inconsistencies between both legacy and new editor:

In the legacy Editor you can set the date without time and it saves the time as 00:00.

https://github.com/woocommerce/google-listings-and-ads/assets/5908855/91cb8e78-dff4-414f-85dd-e2e2d657dfd3

In the new editor, you cannot update til you update the time. If you edit only the date and also another field to allow the update button, the changes are not saved. 

https://github.com/woocommerce/google-listings-and-ads/assets/5908855/0bbcc799-f5d7-47f0-a489-8301b849dac9

eason9487 replied:
Thanks for finding this! Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/2178/commits/b916bd36b5fab085387df6b39ba11999424974e2.

puntope commented on js/src/blocks/product-select-with-text-field/block.json:
```diff
@@ -0,0 +1,40 @@
+	"$schema": "https://schemas.wp.org/trunk/block.json",
+	"apiVersion": 2,
+	"name": "google-listings-and-ads/product-select-with-text-field",
+	"title": "Product select control with text control",
```
💅  In line with the other block. WDYT to keep is as just Fields? 

"Product Select with Text Field"

eason9487 replied:
Adjusted in https://github.com/woocommerce/google-listings-and-ads/commit/776ffb843618a06e08a985a84f56474ed3f49283.

puntope commented on js/src/blocks/product-select-with-text-field/edit.js:
```diff
@@ -0,0 +1,81 @@
+	const handleSelectionChange = ( nextOptionValue ) => {
+		setOptionValue( nextOptionValue );
+
+		if ( nextOptionValue === customInputValue ) {
+			setValue( text );
+		} else {
+			setValue( nextOptionValue );
+		}
```
💅  This can be reduced to:

`setValue( nextOptionValue === customInputValue ? text : nextOptionValue );`

eason9487 replied:
I lead toward keeping it as it is.

puntope commented on js/src/blocks/product-select-with-text-field/edit.js:
```diff
@@ -0,0 +1,81 @@
+
+const FALLBACK_VALUE = '';
+
+export default function Edit( { attributes, context } ) {
```
💅  Can we add a bit of documentation of what `customInputValue` attribute is for?

eason9487 replied:
Added JSDoc for all custom blocks by https://github.com/woocommerce/google-listings-and-ads/pull/2178/commits/d8252b8b9e6214a6223f7fef139a75c267fb4e81.

puntope reviewed (commented):
Thanks @eason9487 for working on this. Overall looking very good. 

I left some comments in the code and a warning in regards the functionallity.

eason9487 committed:
Add an explanation to blocks README for the term "derived value".

Address: https://github.com/woocommerce/google-listings-and-ads/pull/2178#discussion_r1428418544

eason9487 committed:
Correct the inconsistent variable names in js/src/blocks/index.js.

Address: https://github.com/woocommerce/google-listings-and-ads/pull/2178#discussion_r1428424052

eason9487 committed:
Correct the inconsistent titles in each block JSON config.

Address:
- https://github.com/woocommerce/google-listings-and-ads/pull/2178#discussion_r1428426342
- https://github.com/woocommerce/google-listings-and-ads/pull/2178#discussion_r1428454319

eason9487 committed:
Add JSDoc for the custom blocks.

Address: https://github.com/woocommerce/google-listings-and-ads/pull/2178#discussion_r1428462624

eason9487 committed:
Make the date-time custom field block allow the time to be an empty string and fall back to '00:00:00'.

Address: https://github.com/woocommerce/google-listings-and-ads/pull/2178#discussion_r1428452317

eason9487 commented:
@puntope, thanks for the code review. This PR is ready for the new round. Could you take another look?

eason9487 committed:
Merge branch 'feature/support-product-block-editor' into add/product-block-editor-custom-select-with-text-and-date-time

puntope commented on js/src/blocks/product-date-time-field/edit.js:
```diff
@@ -0,0 +1,126 @@
+ * @typedef {import('../types.js').ProductBasicAttributes} ProductBasicAttributes
+ */
+
+async function resolveValidationMessage( inputRef ) {
+	const input = inputRef.current;
+
+	if ( ! input.validity.valid ) {
+		return input.validationMessage;
+	}
+}
```
Why do we need validation for the datetime? I wasn't able to set an invalid date time in the inputs.

eason9487 replied:
I didn't test all browsers, but Chrome and FireFox both allow entering date by keyboard, which may result in an invalid value.

https://github.com/woocommerce/google-listings-and-ads/assets/17420811/f92644b0-e412-4946-b9a9-86ac680ddf13

"請輸入有效的日期" and "請輸入有效的時間" mean "Please enter a valid date" and "Please enter a valid time".

puntope reviewed (approved):
✅  LGTM Thanks for the changes

I left tho a question in regards the validation.
